### PR TITLE
[SCB-1308] improve instance isolation recovery mechanism

### PR DESCRIPTION
**Pull request by yhs0092**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
See details in [SCB-1308](https://issues.apache.org/jira/browse/SCB-1308)

### CODE REVIEW

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java [hunk 1]
```diff
@@ -65,7 +65,19 @@ public ServiceCombServer chooseServer(Invocation invocation) {
...
     }
-    return rule.choose(servers, invocation);
+    ServiceCombServer server = rule.choose(servers, invocation);
+    if (null == server) {
+      return null;
+    }
+    ServiceCombServerStats serverStats = ServiceCombLoadBalancerStats.INSTANCE.getServiceCombServerStats(server);
+    if (ServiceCombServerStatus.ISOLATED == serverStats.getStatus()
+        && serverStats.markStatusAtomically(ServiceCombServerStatus.ISOLATED, ServiceCombServerStatus.TRYING)) {
+      // ensure the status of the chosen isolated instance is changed to TRYING, other isolated instances is still ISOLATED
```
Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java]: for concurrency requests, this mark logic not always work?

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java [hunk 2]
```diff
@@ -65,7 +65,19 @@ public ServiceCombServer chooseServer(Invocation invocation) {
...
       }
     }
-    return rule.choose(servers, invocation);
+    ServiceCombServer server = rule.choose(servers, invocation);
+    if (null == server) {
+      return null;
+    }
+    ServiceCombServerStats serverStats = ServiceCombLoadBalancerStats.INSTANCE.getServiceCombServerStats(server);
+    if (ServiceCombServerStatus.ISOLATED == serverStats.getStatus()
+        && serverStats.markStatusAtomically(ServiceCombServerStatus.ISOLATED, ServiceCombServerStatus.TRYING)) {
```
Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java]: when mark failed, still use the instance?

Comment (wujimin) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java]: and maybe will have ABA problem?

multiple requests run into line 73
one go with success at line 74, and failed very quickly at last, will mark back to ISOLATED
then other requests run to line 74, the markStatusAtomically will still be success, so will try multiple times 

maybe detect the target instance maybe can work by PING mechanism is better?
only PING successed instance allow to TRY in ISOLATED status?
and customer can customize PING mechanism to detect real business status.

Comment (yhs0092) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java]: In `IsolationDiscoveryFilter`, the isolated instances which have passed the `singleTestTime` will be selected into the available instance list to have a try. So here the isolated instances can be used.

Comment (yhs0092) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/LoadBalancer.java]: What about adding a status `LOCKED_FOR_TRYING`? The `ISOLATED` instances are picked out and tansfered to `LOCKED_FOR_TRYING` atomically. And in LoadBalancer, every `LOCKED_FOR_TRYING` instances not selected are transfered back to `ISOLATED`, waiting for the next invocation.

**Comment by WillemJiang** [comment]
Is there any update for the PR?

**Comment by yhs0092** [comment]
Hi, this PR has been refactored. Now the status enumeration in `ServiceCombServerStats` has been removed. And a static `AtomicBoolean` field named `globalAllowIsolatedServerTryingFlag` is added instead to ensure there is not more than 1 instances can stay in trying status concurrently.
Please review again, thanks.

File: handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/IsolationDiscoveryFilter.java [hunk 3]
```diff
@@ -75,6 +82,9 @@ public IsolationDiscoveryFilter() {
       boolean newValue = emptyProtection.get();
       LOGGER.info("{} changed from {} to {}", EMPTY_INSTANCE_PROTECTION, emptyProtection, newValue);
     });
+    if (invocationFinishListenerRegistered.compareAndSet(false, true)) {
```
Comment (liubao68) [inline handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/IsolationDiscoveryFilter.java]: Is this code not used now? I think it's better delete it.

Comment (yhs0092) [reply handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/IsolationDiscoveryFilter.java]: Oh, sorry. It has been fixed.

**Pull request merged**